Improve script builtin checking using schema

zdevito opened the pull request:
* This add aten_schema.h which provides a barebones amount of type and
  argument information about each builtin operator
* emitBuiltinCall is updated to use this information rather than
  aten_dispatch to ensure the operator is correct.
* handling of keyword and position arguments now matches python behavior
* There is no longer a requirement that kwargs be constant or that the
  attributes of an op must be entirely constant or non-constant
* compiler now constructs a non-attributed version of the op first and
  then turns it into the constant-attribute version if all attributes
  are constants.
* default arguments for builtins now work
* SugaredValue::call and similar functions now have SourceRange information
  for their arguments so that error reporting is more accurate

Notes:
* This does not try to merge the builtin checking with python arg parser.
  Given that we will eventually have C10 schema which will replace aten_schema,
  we will eventually have a C++ description of the schema and working of that
  description directly will be the easiest form to understand.
* python function calls and script method calls do not support keyword arguments yet.
  When we add this support we should refactor the handling in tryEmitSchema
  that resolves keywords into a common function.

* default arguments work
* keyword arguments to builtins work (still need to extend to calling python and other script methods)
* much better error reporting for incorrect builtins

Lift any constants to attributes on nodes when possible

Some fail tests

some working tests

zdevito committed:
Improve script builtin checking using schema

* This add aten_schema.h which provides a barebones amount of type and
  argument information about each builtin operator
* emitBuiltinCall is updated to use this information rather than
  aten_dispatch to ensure the operator is correct.
* handling of keyword and position arguments now matches python behavior
* There is no longer a requirement that kwargs be constant or that the
  attributes of an op must be entirely constant or non-constant
* compiler now constructs a non-attributed version of the op first and
  then turns it into the constant-attribute version if all attributes
  are constants.
* default arguments for builtins now work
* SugaredValue::call and similar functions now have SourceRange information
  for their arguments so that error reporting is more accurate

Notes:
* This does not try to merge the builtin checking with python arg parser.
  Given that we will eventually have C10 schema which will replace aten_schema,
  we will eventually have a C++ description of the schema and working of that
  description directly will be the easiest form to understand.
* python function calls and script method calls do not support keyword arguments yet.
  When we add this support we should refactor the handling in tryEmitSchema
  that resolves keywords into a common function.

* default arguments work
* keyword arguments to builtins work (still need to extend to calling python and other script methods)
* much better error reporting for incorrect builtins

Lift any constants to attributes on nodes when possible

Some fail tests

some working tests

zdevito committed:
Add a type field to schema

* This refactors schema to make it usable internally in the compiler as
  the function signatures of script functions as well as for builtin
  operators.
* Adds a List[T] class to better represent the arguments to cat/stack
  as a type rather than with custom checking.

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+
+static std::shared_ptr<SugaredValue> tryEmitSchema(
+  const FunctionSchema& schema,
+  std::stringstream& failure_messages,
```
nit: `ostringstream`

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
-    if(!required)
-      return nullptr;
-    throw ErrorReport(loc) << "unknown builtin op";
+  // fill in position arguments
```
typo: position -> positional

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+      return nullptr;
+    }
+    if(positional_inputs[*idx]) {
+      err() << "argument '" <<  nv.name << "' previously set \n" << nv.loc;
```
"previously set" might be unclear, how about "specified twice"?

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+
+    // we only support lists for builtins, where they must be flattened
+    if(arg.type->kind() == TypeKind::ListType) {
+      auto unpacked = graph->insertNode(graph->createTupleUnpack(v.value));
```
TupleUnpack is a bad name now :confuse

zdevito replied:
where is the confusion here?

apaszke replied:
We're using tuple unpack for lists. It's generally the same as statically retrieving all elements, right?

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+
+  // we successfully matched this schema, construct the node
+
+  // note: we always construct a purely positional nodes here
```
nit: unnecessary `a`

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
     }
-  } else {
-    ensureTensors(loc, inputs);
+    num_outputs = *value;
```
Actually what's `num_outputs` for chunk?

zdevito replied:
It is equal to the number of chunks specified in the chunks=argument. This is why we are checking that this argument is constant here.

apaszke replied:
I was curious what's in that field in the schema

apaszke reviewed (approved):
It was a long one but LGTM! Some minor questions/suggestions.

zdevito committed:
Support kwargs for calls of script methods

This refactors the new schema checking code so that it can be used to
support kwargs and type checking for calls to script methods.

A future commit will be needed to add support for:
* calls to script _functions_ which are currently are GraphExecutors without schema info.
* kwargs to python functions, which will require refactoring python op

zdevito commented on test/test_jit.py:
```diff
@@ -1656,11 +1656,11 @@ def func(x):
         x = torch.rand(10, dtype=torch.float, requires_grad=True)
         y = func(x)
-        y2 = torch.sum(x, dim=0, keepdim=True)
+        y2 = torch.sum(x, dim=0)
```
There is not issue, but previously it was required. I removed it to add a test for default values.

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -239,16 +238,19 @@ def emit_decl(decl):
     } for name in ['sizes', 'strides', 'dim']]
     aten_decls = load_aten_declarations(declarations) + tensor_impl_methods
-    jit_decls = [d for d in aten_decls if is_jit_op(d)]
+    jit_decls = tuple(d for d in aten_decls if is_jit_op(d))
```
I guess I just changed this for debugging. I will switch it back.

zdevito commented on tools/jit/templates/aten_schema.cpp:
```diff
@@ -0,0 +1,101 @@
+    ${names}
+  };
+
+  // default argument values for all ops, represented as using tensors via tensor_as
```
yes

zdevito commented on tools/jit/templates/aten_schema.cpp:
```diff
@@ -0,0 +1,101 @@
+    auto & op = operators[i];
+    schemas.push_back({names[op[0]], getArgumentList(op[1]), getArgumentList(op[2])});
+  }
+  return schemas;
```
yeah, something non-linear was happening when the size of the static initializer got big.

zdevito commented on tools/jit/templates/aten_schema.cpp:
```diff
@@ -0,0 +1,101 @@
+}
+
+static SchemaMap createSchemaMap() {
+  auto schemas = getOperatorSchemas();
```
That is a small omission that causes a big difference in runtime

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -320,7 +323,7 @@ struct GraphFuser {
   // to prepare for fusion and replace uses of n with the new group
   Node * createSingletonFusionGroup(Node * n) {
-    auto group = block->owningGraph()->createFusionGroup(getDevice(n));
+    auto group = block->owningGraph()->createFusionGroup(*getDevice(n));
```
It looks like `.value` throws while `*` doesn't check, so I switched it.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+  // common to see constant lists like [1, 2] passed to attributes
+  bool all_constant = std::all_of(inputs.begin(), inputs.end(), [&](Value* v) {
+    return v->node()->kind() == prim::Constant;
+  });
```
Can you elaborate? I don't understand the issue yet.

apaszke replied:
I thought the function does something else when I was reading it for the first time. Never mind my comment

zdevito commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -103,19 +113,19 @@ struct SimpleValue : public SugaredValue {
+    : name(name), value(std::move(value)) {}
   std::string name;
-  Value* value;
+  at::optional<NamedValue> value;
```
It is self, yes. Commented.

zdevito committed:
PR comments

apaszke commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -20,14 +35,20 @@ std::vector<Value*> Method::emit_call_to(SourceRange loc, Method & callee, Array
+
+  std::stringstream failure_messages;
+  auto all_inputs = tryMatchSchema(
+    callee.schema ? *callee.schema : defaultSchemaFor(callee),
```
When could it happen that a script function we're calling misses schema?

zdevito replied:
Traces don't have schema, so defaultSchemaFor generates reasonable schema for them.

apaszke reviewed (changes requested):
You still haven't changed `createSchemaMap` in a way that would avoid copying all the schemas. Can you please fix that before merging?

yf225 commented:
The CI failure is unrelated to the recent sccache change. I tried the build a couple times with `MAX_JOBS=1` and can confirm that it always fails at compiling `aten_schema.cpp`. For next step I will log the output of sccache invocation and see what’s the problem there.

zdevito commented:
@pytorchbot retest this please

yf225 commented:
@zdevito After inspecting the compilation log, there seems to be a few issues with the generated `aten_schema.cpp` file: 

torch/csrc/jit/generated/aten_schema.cpp: https://gist.githubusercontent.com/yf225/2500ee8d088fa82b8500e298e1c54c0e/raw/5856db317f6128f5c1d1f79e514b96ea55a2a753/generated_aten_schema.cpp

error log: https://gist.githubusercontent.com/yf225/117507fada93ed74f724adcbade04989/raw/8f602e9f96301729716c26fbf0a8aa09e30d121d/sccache_log.txt

I suspect that fixing the "character constant too long for its type" error for all the `odict_keys` will also fix the other errors.

zdevito committed:
Fix python 3

zdevito commented:
This looks like it was a python 2 vs 3 thing. I think it is fixed now. Do you have a sense of why the error messages from the compilation were not reported in the original log?

zdevito committed:
Fix osx

zdevito committed:
Fix libtorch build

yf225 commented:
@zdevito Previously the sccache server was running as a detached daemon and the compilation failure was not sent to stderr. https://github.com/pytorch/pytorch/pull/7555 should fix it.
